### ISSUE TITLE
Immutable state

### ISSUE THREAD

**Issue by arbesfeld** [vuejs/vuex#667]
Hi! We build a library for logging Redux actions / state to a server which relies on the fact that Redux state is immutable, which lets us log the state in the background without having to do a deep clone. 

I see that vuex mutates a state object which makes this impossible. Are there any ways to use an immutable data structure with vuex? 

Thank you! I'm sorry if this is not the right place to ask questions.

**Comment by fnlctrl** [comment]
Hi, thanks for filing this issue, however the issue tracker is reserved for bug reports and feature requests only, and we encourage you to use our dedicated [gitter](https://gitter.im/vuejs/vue) or [forum](http://forum.vuejs.org/) to ask questions or for discussion. Thanks!

============================

Basically no, since vuex/vue's reactivity relies on mutability. However, if you don't need the reactivity, simply call `Object.freeze(obj)` before storing it in the store, and it will remain immutable.

If you rely on reactivity (i.e. the state is used in rendering view), there's a simple technique to make the object that holds the immutable state reactive, and use it to trigger view updates.

For example:
```js
{
  state: {
    reactiveHolder: {
      immutableState: Object.freeze({...state object...})
      ...
    }
  }
}
```

Now, when using the `immutableState` in view, instead of accessing the object itself, use `reactiveHolder[key]` where key is `'immutableState'`. And when updating `immutableState`, use `reactiveHolder[key] = Object.freeze({...state object...})`. 
In other words, use a reference(`holder[key]`) to the immutable object instead of the object itself when using in view.
